lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.cpp

```diff
@@ -2584,37 +2586,104 @@ lldb::addr_t SwiftLanguageRuntime::GetAsyncContext(RegisterContext *regctx) {
-/// `with_double_deref` is true. This is only valid for x86_64 or aarch64.
-llvm::ArrayRef<uint8_t>
-GetAsyncRegFromFramePointerDWARFExpr(llvm::Triple::ArchType triple,
-                                     bool with_double_deref) {
-  assert(triple == llvm::Triple::x86_64 || triple == llvm::Triple::aarch64);
-
-  // These expressions must have static storage, due to how UnwindPlan::Row
-  // works.
-  static const uint8_t g_cfa_dwarf_expression_x86_64[] = {
-      llvm::dwarf::DW_OP_breg6, // DW_OP_breg6, register 6 == rbp
-      0x78,                     //    sleb128 -8 (ptrsize)
-      llvm::dwarf::DW_OP_deref,
-      llvm::dwarf::DW_OP_deref,
-  };
-  static const uint8_t g_cfa_dwarf_expression_arm64[] = {
-      llvm::dwarf::DW_OP_breg29, // DW_OP_breg29, register 29 == fp
-      0x78,                      //    sleb128 -8 (ptrsize)
-      llvm::dwarf::DW_OP_deref,
-      llvm::dwarf::DW_OP_deref,
-  };
+/// Functional wrapper to read a register as an address.
+std::optional<addr_t> ReadRegisterAsAddress(RegisterContext &regctx,
+                                            unsigned regnum) {
+  auto reg = regctx.ReadRegisterAsUnsigned(regnum, LLDB_INVALID_ADDRESS);
+  if (reg != LLDB_INVALID_ADDRESS)
+    return reg;
+  return {};
+}
+
+/// Functional wrapper to read a pointer from process memory at `addr +
+/// offset`.
+std::optional<addr_t>
+ReadPtrFromAddr(Process &process, std::optional<addr_t> addr, int offset = 0) {
+  if (!addr)
+    return {};
+  Status error;
+  addr_t ptr = process.ReadPointerFromMemory(*addr + offset, error);
+  if (ptr != LLDB_INVALID_ADDRESS)
+    return ptr;
+  return {};
+}
+
+/// Converts the abstract location of UnwindPlan::Row::FAValue into a concrete
+/// address. This is a simplified version of the methods in
+/// RegisterContextUnwind, since plumbing access to those here would be
+/// challenging.
+std::optional<addr_t> GetCFAAddr(Process &process, RegisterContext &regctx,
+                                 UnwindPlan::Row::FAValue cfa_loc) {
+  if (cfa_loc.IsRegisterPlusOffset()) {
+    unsigned regnum = cfa_loc.GetRegisterNumber();
+    if (std::optional<addr_t> regvalue = ReadRegisterAsAddress(regctx, regnum))
+      return *regvalue + cfa_loc.GetOffset();
+  }
+  return {};
+}
+
+/// Attempts to use UnwindPlans that inspect assembly to recover the entry value
+/// of the async register. This is a simplified version of the methods in
+/// RegisterContextUnwind, since plumbing access to those here would be
+/// challenging.
+std::optional<addr_t>
+ReadAsyncRegFromUnwind(SymbolContext &sc, Process &process, Address pc,
+                       Address func_start_addr, RegisterContext &regctx,
+                       AsyncUnwindRegisterNumbers regnums) {
+  FuncUnwindersSP unwinders =
+      pc.GetModule()->GetUnwindTable().GetFuncUnwindersContainingAddress(pc,
+                                                                         sc);
+  if (!unwinders)
+    return {};
```

Comment:
Is there a benefit to converting this to Expected and returning more specific StringErrors here? Would that help with debugging if something goes wrong?

Reply:
Or does the failure just mean "this is not an async function"?

Reply:
failure here would mean our unwinders were not able to inspect the code for some reason. Let me try to add a separate commit that uses a mix of expected + logging to see what it would look like, and then we can decide if it's worth the trouble